lib/ts/controllers/s-popover.ts

```diff
@@ -315,6 +324,177 @@ namespace Stacks {
+     *                If the popover does not have a parent element, it will be inserted as a immediately after the reference element.
+     * @param options an optional collection of options to use when configuring the popover.
+     */
+     export function attachPopover(element: Element, popover: Element | string, options?: PopoverOptions)
+     {
+        const { referenceElement, popover: existingPopover } = getPopover(element);
+
+        if (existingPopover) {
+            throw `element already has popover with id="${existingPopover.id}"`
+        }
+
+        if (!referenceElement) {
+            throw `element has invalid data-s-popover-reference-selector attribute`
+        }
+
+        if (typeof popover === 'string') {
+            const elements = document.createRange().createContextualFragment(popover).children;
+            if (elements.length !== 1) {
+                throw "popover should contain a single element";
+            }
+            popover = elements[0];
+        }
+
+        const existingId = referenceElement.getAttribute("aria-controls");
+        var popoverId = popover.id;
+
+        if (!popover.classList.contains('s-popover')) {
+            throw `popover should have the "s-popover" class but had class="${popover.className}"`;
+        }
+
+        if (existingId && existingId !== popoverId) {
+            throw `element has aria-controls="${existingId}" but popover has id="${popoverId}"`;
+        }
+
+        if (!popoverId) {
+            popoverId = "--stacks-s-popover-" + Math.random().toString(36).substring(2, 10);
+            popover.id = popoverId;
+        }
+
+        if (!existingId) {
+            referenceElement.setAttribute("aria-controls", popoverId);
+        }
+
+        if (!popover.parentElement && element.parentElement) {
+            referenceElement.insertAdjacentElement("afterend", popover);
+        }
+
+        element.setAttribute("data-controller", "s-popover");
+
+        if (options) {
+            if (options.toggleOnClick) {
+                referenceElement.setAttribute("data-action", "click->s-popover#toggle");
+            }
+            if (options.placement) {
+                element.setAttribute("data-s-popover-placement", options.placement);
+            }
+            if (options.autoShow) {
+                element.setAttribute("data-s-popover-auto-show", "true");
+            }
+        }
+    }
+
+    /**
+     * Removes the popover controller from an element and removes the popover from the DOM.
+     * @param element the element that has the `data-controller="s-popover"` attribute.
+     * @returns The popover that was attached to the element.
+     */
+    export function detachPopover(element: Element) {
+        const { isPopover, controller, referenceElement, popover } = getPopover(element);
+
+        // Hide the popover so its events fire.
+        if (controller) { controller.hide(); }
```

Comment:
```suggestion
        controller?.hide();
```

Maybe shorthand here?